File: filebridge-py/src/filebridge/stream.py
"""Streaming protocol and encryption primitives for the Filebridge wire format."""

from __future__ import annotations

import base64
import hashlib
import hmac as _hmac
import json
import struct
from typing import Any, NamedTuple

import zstandard
from cryptography.exceptions import InvalidTag
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.ciphers.aead import ChaCha20Poly1305
from cryptography.hazmat.primitives.kdf.hkdf import HKDFExpand

from .exceptions import FileBridgeError, IsDirectoryError

CHUNK_SIZE = 64 * 1024
"""Default chunk size (64 KiB) for stream framing and buffered writes."""


class StreamError(Exception):
    """Raised on stream protocol or cryptographic errors."""


def _hkdf_extract(salt: bytes, ikm: bytes) -> bytes:
    """HKDF-Extract step: PRK = HMAC-SHA256(salt, ikm)."""
    return _hmac.new(salt, ikm, hashlib.sha256).digest()


def _hkdf_expand(prk: bytes, info: bytes, length: int) -> bytes:
    """HKDF-Expand step using cryptography library."""
    return HKDFExpand(algorithm=hashes.SHA256(), length=length, info=info).derive(prk)


def _derive_stream_key_nonce(token: str, iv_hex: str) -> tuple[bytes, bytes]:
    """Derive key and nonce base for stream AEAD encryption/decryption."""
    if not iv_hex:
        raise ValueError("iv_hex must not be empty")
    prk = _hkdf_extract(iv_hex.encode(), token.encode())
    key = _hkdf_expand(prk, b"filebridge-stream-key", 32)
    nonce_base = _hkdf_expand(prk, b"filebridge-stream-nonce", 12)
    return key, nonce_base


def _derive_json_key_nonce(token: str, iv_hex: str) -> tuple[bytes, bytes]:
    """Derive key and nonce for JSON response encryption/decryption."""
    if not iv_hex:
        raise ValueError("iv_hex must not be empty")
    prk = _hkdf_extract(iv_hex.encode(), token.encode())
    key = _hkdf_expand(prk, b"filebridge-json-key", 32)
    nonce = _hkdf_expand(prk, b"filebridge-json-nonce", 12)
    return key, nonce


class StreamAead:
    """ChaCha20-Poly1305 AEAD for encrypting/decrypting stream chunks."""

    def __init__(self, token: str, iv_hex: str):
        key, nonce_base = _derive_stream_key_nonce(token, iv_hex)
        self.cipher = ChaCha20Poly1305(key)
        self.nonce_base = bytearray(nonce_base)
        self.counter = 0

    def _current_nonce(self) -> bytes:
        nonce_bytes = bytearray(self.nonce_base)
        counter_bytes = struct.pack(">Q", self.counter)
        for i in range(8):
            nonce_bytes[4 + i] ^= counter_bytes[i]
        return bytes(nonce_bytes)

    def encrypt(self, data: bytes) -> bytes:
        nonce = self._current_nonce()
        ciphertext = self.cipher.encrypt(nonce, data, None)
        self.counter += 1
        return ciphertext

    def decrypt(self, data: bytes) -> bytes:
        nonce = self._current_nonce()
        try:
            plaintext = self.cipher.decrypt(nonce, data, None)
            self.counter += 1
            return plaintext
        except InvalidTag as e:
            raise StreamError(f"Decryption failed: {e}")

    def finalize(self) -> str:
        final_block = self.encrypt(b"")
        return final_block.hex()

    def verify_stop(self, hex_sig: str):
        try:
            final_block = bytes.fromhex(hex_sig)
            self.decrypt(final_block)
        except (ValueError, StreamError) as e:
            raise StreamError(f"Invalid stop signature: {e}")


def encode_meta(payload: bytes) -> bytes:
    frame = bytearray(b"META")
    frame.extend(struct.pack(">I", len(payload)))
    frame.extend(payload)
    return bytes(frame)


def encode_data(payload: bytes) -> bytes:
    frame = bytearray(b"DATA")
    frame.extend(struct.pack(">I", len(payload)))
    frame.extend(payload)
    return bytes(frame)


def encode_stop(signature: str | None = None) -> bytes:
    frame = bytearray(b"STOP")
    if signature:
        sig_bytes = signature.encode()
        frame.extend(struct.pack(">I", len(sig_bytes)))
        frame.extend(sig_bytes)
    else:
        frame.extend(struct.pack(">I", 0))
    return bytes(frame)


def _derive_block_nonce(token: str, iv_hex: str) -> tuple[ChaCha20Poly1305, bytes]:
    """Shared key+nonce derivation for single-block JSON encryption."""
    key, nonce = _derive_json_key_nonce(token, iv_hex)
    cipher = ChaCha20Poly1305(key)
    return cipher, nonce


def encrypt_json_response(token: str, iv_hex: str, json_bytes: bytes) -> str:
    """Compress, encrypt and base64-encode JSON bytes for a token-protected response."""
    compressed = zstandard.ZstdCompressor(level=3).compress(json_bytes)
    cipher, nonce = _derive_block_nonce(token, iv_hex)
    ciphertext = cipher.encrypt(nonce, compressed, None)
    return base64.b64encode(ciphertext).decode("ascii")


def decrypt_json_response(token: str, iv_hex: str, encoded: str) -> bytes:
    """Base64-decode, decrypt and decompress an encrypted JSON response."""
    data = base64.b64decode(encoded)
    cipher, nonce = _derive_block_nonce(token, iv_hex)
    try:
        compressed = cipher.decrypt(nonce, data, None)
    except InvalidTag as e:
        raise StreamError(f"JSON response decryption failed: {e}")
    return zstandard.ZstdDecompressor().decompress(compressed)


class StreamFrame(NamedTuple):
    """A decoded stream frame."""

    tag: str
    signature: str | None
    payload: bytes


class StreamDecoder:
    """Incremental decoder for the filebridge binary stream protocol.

    Feed data with ``push()``, then call ``next_frame()`` repeatedly
    to extract complete frames.
    """

    def __init__(self):
        self.buffer = bytearray()
        self.offset = 0

    def push(self, chunk: bytes):
        """Append raw bytes to the internal buffer."""
        self.buffer.extend(chunk)

    def next_frame(self) -> StreamFrame | None:
        """Return the next complete frame, or ``None`` if more data is needed."""
        rem_len = len(self.buffer) - self.offset
        if rem_len < 8:
            return None

        tag = bytes(self.buffer[self.offset : self.offset + 4]).decode("ascii")
        length = struct.unpack(">I", self.buffer[self.offset + 4 : self.offset + 8])[0]

        if rem_len < 8 + length:
            return None

        payload = bytes(self.buffer[self.offset + 8 : self.offset + 8 + length])

        # Advance buffer offset
        self.offset += 8 + length

        # Periodically compact buffer to release memory if we've processed a lot
        if self.offset > 1024 * 1024:
            del self.buffer[: self.offset]
            self.offset = 0

        sig_str = None
        if tag == "STOP":
            if length > 0:
                sig_str = payload.decode("ascii")

        return StreamFrame(tag=tag, signature=sig_str, payload=payload)


# ---------------------------------------------------------------------------
# Higher-level operations that combine crypto primitives with the wire format
# ---------------------------------------------------------------------------


def build_encrypted_envelope(
    token: str,
    sig: str,
    path: str,
    offset: int | None = None,
    length: int | None = None,
    extensive: bool = False,
) -> str:
    """Build an encrypted request envelope containing path and params."""
    envelope: dict[str, Any] = {"path": path.lstrip("/")}
    if offset is not None:
        envelope["offset"] = offset
    if length is not None:
        envelope["length"] = length
    if extensive:
        envelope["extensive"] = True
    json_bytes = json.dumps(envelope, separators=(",", ":")).encode()
    return encrypt_json_response(token, sig, json_bytes)


def parse_json_response(token: str | None, signature: str | None, body: bytes) -> dict:
    """Parse a JSON response body, decrypting it first when token+signature are present."""
    parsed = json.loads(body)
    if token and signature:
        message = parsed.get("message")
        if message is None:
            raise FileBridgeError("Missing 'message' field in encrypted response")
        try:
            json_bytes = decrypt_json_response(token, signature, message)
        except StreamError as e:
            raise FileBridgeError(f"JSON response decryption failed: {e}")
        return json.loads(json_bytes)
    return parsed


def build_encrypted_write_body(
    token: str,
    sig: str,
    data: bytes,
    path: str | None = None,
    offset: int | None = None,
) -> bytes:
    """Pack `data` into signed stream frames (ChaCha20Poly1305).

    When *path* is given, a META frame with the encrypted envelope is prepended
    (token-mode: path not in URL).
    """
    buf = bytearray()
    if path is not None:
        envelope = build_encrypted_envelope(token, sig, path, offset)
        buf.extend(encode_meta(envelope.encode()))
    aead = StreamAead(token, sig)
    for i in range(0, len(data), CHUNK_SIZE):
        buf.extend(encode_data(aead.encrypt(data[i : i + CHUNK_SIZE])))
    buf.extend(encode_stop(aead.finalize()))
    return bytes(buf)


def decode_verified_stream_content(token: str, signature: str, content: bytes) -> bytes:
    """Decrypt and verify a complete stream response body."""
    decoder = StreamDecoder()
    aead = StreamAead(token, signature)
    decoder.push(content)

    result_bytes = bytearray()
    while True:
        frame = decoder.next_frame()
        if not frame:
            break
        tag, sig_str, payload = frame
        if tag == "DATA":
            try:
                result_bytes.extend(aead.decrypt(payload))
            except StreamError:
                raise FileBridgeError("Chunk Authenticated Decryption failed")
        elif tag == "STOP":
            if not sig_str:
                raise FileBridgeError("Stop frame missing signature")
            try:
                aead.verify_stop(sig_str)
            except StreamError:
                raise FileBridgeError("Stop signature mismatch")
            break
    return bytes(result_bytes)


def decode_read_response(
    token: str | None,
    content_type: str,
    content: bytes,
    sig: str | None,
    path: str,
) -> bytes:
    """Evaluate Content-Type and return decoded payload bytes.

    Raises IsDirectoryError for directory JSON responses, FileBridgeError
    for missing signature in stream mode.
    """
    if "application/json" in content_type:
        data = parse_json_response(token, sig if token else None, content)
        if "items" in data:
            raise IsDirectoryError(f"{path} is a directory")

    if "application/vnd.filebridge.stream" in content_type:
        if not sig or not token:
            raise FileBridgeError("Missing signature for stream verification")
        return decode_verified_stream_content(token, sig, content)

    return content

File: filebridge/src/stream.rs
//! Streaming protocol and encryption primitives for the Filebridge wire format.

use bytes::{Buf, Bytes, BytesMut};
use chacha20poly1305::{
    aead::{AeadInPlace, KeyInit},
    ChaCha20Poly1305, Key, Nonce,
};
use hkdf::Hkdf;
use sha2::Sha256;
use thiserror::Error;

#[derive(Debug, Error)]
pub enum StreamError {
    #[error("Invalid stream frame tag: {0:?}")]
    InvalidTag([u8; 4]),
    #[error("Cryptographic error: {0}")]
    CryptoError(String),
    #[error("HKDF error: {0}")]
    Hkdf(&'static str),
}

pub enum StreamFrame {
    Meta { payload: Bytes },
    Data { payload: Bytes },
    Stop { signature: Option<String> }, // signature is the final AEAD tag hex
}

pub fn encode_meta(payload: &[u8]) -> Vec<u8> {
    let mut frame = Vec::with_capacity(8 + payload.len());
    frame.extend_from_slice(b"META");
    frame.extend_from_slice(&(payload.len() as u32).to_be_bytes());
    frame.extend_from_slice(payload);
    frame
}

pub fn encode_data(payload: &[u8]) -> Vec<u8> {
    let mut frame = Vec::with_capacity(8 + payload.len());
    frame.extend_from_slice(b"DATA");
    frame.extend_from_slice(&(payload.len() as u32).to_be_bytes());
    frame.extend_from_slice(payload);
    frame
}

pub fn encode_stop(signature: Option<&str>) -> Vec<u8> {
    let mut frame = Vec::new();
    frame.extend_from_slice(b"STOP");
    if let Some(sig) = signature {
        let sig_bytes = sig.as_bytes();
        frame.extend_from_slice(&(sig_bytes.len() as u32).to_be_bytes());
        frame.extend_from_slice(sig_bytes);
    } else {
        frame.extend_from_slice(&(0u32).to_be_bytes());
    }
    frame
}

pub struct StreamDecoder {
    buffer: BytesMut,
}

impl Default for StreamDecoder {
    fn default() -> Self {
        Self::new()
    }
}

impl StreamDecoder {
    pub fn new() -> Self {
        Self {
            buffer: BytesMut::new(),
        }
    }

    pub fn push(&mut self, chunk: &[u8]) {
        self.buffer.extend_from_slice(chunk);
    }

    /// Return any remaining unprocessed bytes in the buffer.
    pub fn remaining(&self) -> &[u8] {
        &self.buffer
    }

    pub fn next_frame(&mut self) -> Result<Option<StreamFrame>, StreamError> {
        if self.buffer.len() < 8 {
            return Ok(None);
        }

        let mut tag = [0u8; 4];
        tag.copy_from_slice(&self.buffer[0..4]);

        let length = u32::from_be_bytes([
            self.buffer[4],
            self.buffer[5],
            self.buffer[6],
            self.buffer[7],
        ]) as usize;

        if self.buffer.len() < 8 + length {
            return Ok(None);
        }

        self.buffer.advance(8);
        let payload = self.buffer.split_to(length);

        match &tag {
            b"META" => Ok(Some(StreamFrame::Meta {
                payload: payload.freeze(),
            })),
            b"DATA" => Ok(Some(StreamFrame::Data {
                payload: payload.freeze(),
            })),
            b"STOP" => {
                let sig = if length > 0 {
                    Some(
                        String::from_utf8(payload.to_vec())
                            .map_err(|_| StreamError::CryptoError("Invalid UTF-8 in STOP frame signature".to_string()))?,
                    )
                } else {
                    None
                };
                Ok(Some(StreamFrame::Stop { signature: sig }))
            }
            _ => Err(StreamError::InvalidTag(tag)),
        }
    }
}

/// Derive key and nonce base for stream AEAD encryption/decryption.
fn derive_stream_key_nonce(token: &str, iv_hex: &str) -> Result<([u8; 32], [u8; 12]), StreamError> {
    if iv_hex.is_empty() {
        return Err(StreamError::Hkdf("iv_hex must not be empty"));
    }
    let hk = Hkdf::<Sha256>::new(Some(iv_hex.as_bytes()), token.as_bytes());
    let mut key = [0u8; 32];
    hk.expand(b"filebridge-stream-key", &mut key)
        .map_err(|_| StreamError::Hkdf("expand failed for stream key"))?;
    let mut nonce_base = [0u8; 12];
    hk.expand(b"filebridge-stream-nonce", &mut nonce_base)
        .map_err(|_| StreamError::Hkdf("expand failed for stream nonce"))?;
    Ok((key, nonce_base))
}

/// Derive key and nonce for JSON response encryption/decryption.
/// Uses different info strings than stream AEAD for purpose separation.
fn derive_json_key_nonce(token: &str, iv_hex: &str) -> Result<([u8; 32], [u8; 12]), StreamError> {
    if iv_hex.is_empty() {
        return Err(StreamError::Hkdf("iv_hex must not be empty"));
    }
    let hk = Hkdf::<Sha256>::new(Some(iv_hex.as_bytes()), token.as_bytes());
    let mut key = [0u8; 32];
    hk.expand(b"filebridge-json-key", &mut key)
        .map_err(|_| StreamError::Hkdf("expand failed for JSON key"))?;
    let mut nonce = [0u8; 12];
    hk.expand(b"filebridge-json-nonce", &mut nonce)
        .map_err(|_| StreamError::Hkdf("expand failed for JSON nonce"))?;
    Ok((key, nonce))
}

pub struct StreamAead {
    cipher: ChaCha20Poly1305,
    nonce_base: [u8; 12],
    counter: u64,
}

impl StreamAead {
    pub fn new(token: &str, iv_hex: &str) -> Result<Self, StreamError> {
        let (key, nonce_base) = derive_stream_key_nonce(token, iv_hex)?;
        let cipher = ChaCha20Poly1305::new(Key::from_slice(&key));

        Ok(Self {
            cipher,
            nonce_base,
            counter: 0,
        })
    }

    fn current_nonce(&self) -> [u8; 12] {
        let mut nonce_bytes = self.nonce_base;
        let counter_bytes = self.counter.to_be_bytes(); // 8 bytes
        for i in 0..8 {
            nonce_bytes[4 + i] ^= counter_bytes[i];
        }
        nonce_bytes
    }

    pub fn encrypt(&mut self, data: &mut Vec<u8>) -> Result<(), StreamError> {
        let binding = self.current_nonce();
        let nonce = Nonce::from_slice(&binding);
        self.cipher
            .encrypt_in_place(nonce, b"", data)
            .map_err(|e| StreamError::CryptoError(format!("Encryption failed: {}", e)))?;
        self.counter += 1;
        Ok(())
    }

    pub fn decrypt(&mut self, data: &mut Vec<u8>) -> Result<(), StreamError> {
        let binding = self.current_nonce();
        let nonce = Nonce::from_slice(&binding);
        self.cipher
            .decrypt_in_place(nonce, b"", data)
            .map_err(|e| StreamError::CryptoError(format!("Decryption failed: {}", e)))?;
        self.counter += 1;
        Ok(())
    }

    pub fn finalize(&mut self) -> Result<String, StreamError> {
        let mut final_block = Vec::new();
        self.encrypt(&mut final_block)?;
        Ok(hex::encode(final_block))
    }

    pub fn verify_stop(&mut self, hex_sig: &str) -> Result<(), StreamError> {
        let mut final_block = hex::decode(hex_sig)
            .map_err(|e| StreamError::CryptoError(format!("Invalid hex in stop signature: {}", e)))?;
        self.decrypt(&mut final_block)?;
        Ok(())
    }
}

/// Encrypt JSON bytes for a response: compress → encrypt → base64.
/// Uses the same key/nonce derivation as `StreamAead` with counter=0.
pub fn encrypt_json_response(token: &str, iv_hex: &str, json_bytes: &[u8]) -> Result<String, StreamError> {
    use base64::Engine as _;

    let mut encoder = zstd::Encoder::new(Vec::new(), 3)
        .map_err(|e| StreamError::CryptoError(format!("Compression failed: {}", e)))?;
    encoder
        .include_contentsize(true)
        .map_err(|e| StreamError::CryptoError(format!("Compression setup failed: {}", e)))?;
    encoder
        .set_pledged_src_size(Some(json_bytes.len() as u64))
        .map_err(|e| StreamError::CryptoError(format!("Compression setup failed: {}", e)))?;
    {
        use std::io::Write;
        encoder
            .write_all(json_bytes)
            .map_err(|e| StreamError::CryptoError(format!("Compression write failed: {}", e)))?;
    }
    let compressed = encoder
        .finish()
        .map_err(|e| StreamError::CryptoError(format!("Compression finish failed: {}", e)))?;

    let (key, nonce_bytes) = derive_json_key_nonce(token, iv_hex)?;
    let cipher = ChaCha20Poly1305::new(Key::from_slice(&key));
    let nonce = Nonce::from_slice(&nonce_bytes);

    let mut data = compressed;
    cipher
        .encrypt_in_place(nonce, b"", &mut data)
        .map_err(|e| StreamError::CryptoError(format!("Encryption failed: {}", e)))?;

    Ok(base64::engine::general_purpose::STANDARD.encode(&data))
}

/// Decrypt a JSON response: base64-decode → decrypt → decompress.
pub fn decrypt_json_response(token: &str, iv_hex: &str, encoded: &str) -> Result<Vec<u8>, StreamError> {
    use base64::Engine as _;

    let mut data = base64::engine::general_purpose::STANDARD
        .decode(encoded)
        .map_err(|e| StreamError::CryptoError(format!("Base64 decode failed: {}", e)))?;

    let (key, nonce_bytes) = derive_json_key_nonce(token, iv_hex)?;
    let cipher = ChaCha20Poly1305::new(Key::from_slice(&key));
    let nonce = Nonce::from_slice(&nonce_bytes);

    cipher
        .decrypt_in_place(nonce, b"", &mut data)
        .map_err(|e| StreamError::CryptoError(format!("Decryption failed: {}", e)))?;

    zstd::decode_all(data.as_slice())
        .map_err(|e| StreamError::CryptoError(format!("Decompression failed: {}", e)))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_json_response_roundtrip() {
        let token = "test-secret";
        let iv_hex = "abcdef1234567890abcdef1234567890abcdef1234567890abcdef1234567890";
        let json = br#"{"items":[{"name":"foo.txt","is_dir":false}]}"#;

        let encoded = encrypt_json_response(token, iv_hex, json).unwrap();
        let decoded = decrypt_json_response(token, iv_hex, &encoded).unwrap();
        assert_eq!(decoded, json);
    }

    #[test]
    fn test_json_response_wrong_token_fails() {
        let token = "test-secret";
        let iv_hex = "abcdef1234567890abcdef1234567890abcdef1234567890abcdef1234567890";
        let json = br#"{"name":"test.txt","is_dir":false}"#;

        let encoded = encrypt_json_response(token, iv_hex, json).unwrap();
        assert!(decrypt_json_response("wrong-token", iv_hex, &encoded).is_err());
    }

    #[test]
    fn test_hkdf_cross_language_vectors() {
        // These values must match the Python implementation exactly
        let token = "test-secret";
        let iv_hex = "abcdef1234567890abcdef1234567890abcdef1234567890abcdef1234567890";

        let (sk, sn) = derive_stream_key_nonce(token, iv_hex).unwrap();
        assert_eq!(hex::encode(sk), "4e1f082bb3d869ec18ee2f8a06fcf9cd8741f2d847fd443a1b6a0d761a7405e7");
        assert_eq!(hex::encode(sn), "fd55d161ad566053ff2d2a0d");

        let (jk, jn) = derive_json_key_nonce(token, iv_hex).unwrap();
        assert_eq!(hex::encode(jk), "5325c7eb06e53ce49800fb593a2aadc6b77453a03c0e1e99c843ef094deb3f42");
        assert_eq!(hex::encode(jn), "666c2f925009050d9f0c4069");
    }

    #[test]
    fn test_stream_aead_roundtrip() {
        let token = "test-secret";
        let iv_hex = "abcdef1234567890";

        let mut enc = StreamAead::new(token, iv_hex).unwrap();
        let mut chunk1 = b"hello world".to_vec();
        enc.encrypt(&mut chunk1).unwrap();
        let mut chunk2 = b"second chunk".to_vec();
        enc.encrypt(&mut chunk2).unwrap();
        let stop = enc.finalize().unwrap();

        let mut dec = StreamAead::new(token, iv_hex).unwrap();
        dec.decrypt(&mut chunk1).unwrap();
        assert_eq!(chunk1, b"hello world");
        dec.decrypt(&mut chunk2).unwrap();
        assert_eq!(chunk2, b"second chunk");
        dec.verify_stop(&stop).unwrap();
    }
}
